runtime/vm/mz64/UpcallThunkGen.cpp

```diff
@@ -247,32 +503,44 @@ createUpcallThunk(J9UpcallMetaData *metaData)
-    /* The index for the return type in the signature array */
-    I_32 lastSigIdx = (I_32)(nativeSig->numSigs - 1);
+	J9UpcallSigType *sigArray = nativeSig->sigArray;
+	/* The index for the return type in the signature array */
+	const I_32 lastSigIdx = (I_32)(nativeSig->numSigs - 1);
+	U_8 returnType = sigArray[lastSigIdx].type;
+	U_32 returnValueSize = sigArray[lastSigIdx].sizeInByte;
 
-    Assert_VM_true((argIdx >= 0) && (argIdx < lastSigIdx));
+	Assert_VM_true((argIdx >= 0) && (argIdx < lastSigIdx));
 
-    I_32 stackSlotCount = 0;
+	I_32 stackSlotCount = 0;
+	if (((returnType & J9_FFI_UPCALL_SIG_TYPE_MASK) == J9_FFI_UPCALL_SIG_TYPE_STRUCT) && returnValueSize > 24) {
+		// In the case where the return value's type is a struct, we must account for the hidden argument.
+		// All structs <= 24 will be stored in GPRs or FPRs. The rest are stored in a buffer in memory. The buffer's address is the first double
+		// word in the argument area on the stack. Since this is considered a "hidden" argument, we must skip over this slot if it is present.
+		stackSlotCount++;
+	}
 
-    /* Loop through the arguments. All arguments are in the caller frame, so this is simple. */
-    for (I_32 i = 0; i < argIdx; i++) {
-        /* Testing this argument */
-        switch (sigArray[i].type & J9_FFI_UPCALL_SIG_TYPE_MASK) {
-            case J9_FFI_UPCALL_SIG_TYPE_CHAR:    /* Fall through */
-            case J9_FFI_UPCALL_SIG_TYPE_SHORT:   /* Fall through */
-            case J9_FFI_UPCALL_SIG_TYPE_INT32:   /* Fall through */
-            case J9_FFI_UPCALL_SIG_TYPE_POINTER: /* Fall through */
-            case J9_FFI_UPCALL_SIG_TYPE_INT64:   /* Fall through */
-            case J9_FFI_UPCALL_SIG_TYPE_FLOAT:   /* Fall through */
-            case J9_FFI_UPCALL_SIG_TYPE_DOUBLE:
-                stackSlotCount += 1;
-                break;
-            default:
-                Assert_VM_unreachable();
-        }
-    }
-    return (void *)((char *)argListPtr + (stackSlotCount * 8));
+	/* Loop through the arguments. All arguments are in the caller frame, so this is simple. */
+	for (I_32 i = 0; i < argIdx; i++) {
+		U_32 argSizeInBytes = sigArray[i].sizeInByte;
+		/* Testing this argument */
+		switch (sigArray[i].type & J9_FFI_UPCALL_SIG_TYPE_MASK) {
+			case J9_FFI_UPCALL_SIG_TYPE_CHAR:    /* Fall through */
+			case J9_FFI_UPCALL_SIG_TYPE_SHORT:   /* Fall through */
+			case J9_FFI_UPCALL_SIG_TYPE_INT32:   /* Fall through */
+			case J9_FFI_UPCALL_SIG_TYPE_POINTER: /* Fall through */
+			case J9_FFI_UPCALL_SIG_TYPE_INT64:   /* Fall through */
+			case J9_FFI_UPCALL_SIG_TYPE_FLOAT:   /* Fall through */
+			case J9_FFI_UPCALL_SIG_TYPE_DOUBLE:
+				stackSlotCount += 1;
+				break;
+			case J9_FFI_UPCALL_SIG_TYPE_STRUCT:
+				stackSlotCount += argSizeInBytes % 8 == 0 ? argSizeInBytes / 8 : argSizeInBytes / 8 + 1;
+				break;
+			default:
+				Assert_VM_unreachable();
+		}
+	}
+	return (void *)((char *)argListPtr + (stackSlotCount * 8));
```

Comment:
Instead of `char*` , can we use `U_8 *` ?

Reply:
Done: b6cac56c59ac275b1cd08ba52da968d2a7cb5e94